library/std/src/collections/hash/map.rs

```diff
@@ -2501,6 +2502,58 @@ where
 }
 
 impl<'a, K, V> Entry<'a, K, V> {
+    /// Inserts `value` if the entry is vacant, otherwise merges the existing
+    /// value with `value`, using the given function `f`.
+    ///
+    /// If a panic occurs in the function, the entry will be removed.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(hashmap_entry_merge)]
+    ///
+    /// use std::collections::HashMap;
+    /// use std::ops::Add;
+    ///
+    /// let mut word_counts: HashMap<&str, u32> = HashMap::new();
+    /// word_counts.entry("very").merge(1, Add::add);
+    /// word_counts.entry("very").merge(1, Add::add);
+    /// word_counts.entry("wonderful").merge(1, Add::add);
+    ///
+    /// assert_eq!(word_counts["very"], 2);
+    /// assert_eq!(word_counts["wonderful"], 1);
+    /// ```
+    #[unstable(feature = "hashmap_entry_merge", issue = "none")]
+    pub fn merge(self, value: V, f: impl FnOnce(V, V) -> V) -> &'a mut V {
+        struct Guard<'a, K, V> {
+            copied: Option<OccupiedEntry<'a, K, V>>,
+        }
+        impl<K, V> Drop for Guard<'_, K, V> {
+            fn drop(&mut self) {
+                if let Some(entry) = self.copied.take() {
+                    mem::forget(entry.remove());
```

Comment:
This is technically creating a value of type `V` that's invalid (because already dropped). It might be fine, but to be 100% sure it would be nice to have a way to "forget" an entry.

Reply:
There is [`RawTable::erase_no_drop`](https://docs.rs/hashbrown/latest/hashbrown/raw/struct.RawTable.html#method.erase_no_drop) in hashbrown but I don't think I can access that method from here, unless a new function were added to `RustcOccupiedEntry`. The extra read here is probably optimized out though?

Reply:
> There is [RawTable::erase_no_drop](https://docs.rs/hashbrown/latest/hashbrown/raw/struct.RawTable.html#method.erase_no_drop) in hashbrown but I don't think I can access that method from here, unless a new function were added to RustcOccupiedEntry.

Yeah, that was my idea.

> The extra read here is probably optimized out though?

I'm not worried about the extra work, but about whether this is UB or not. With the `entry.remove()` you're technically creating an invalid value, even though it's immediately forgotten. See [this playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bdbb0fcd21cf622a853228ae0b1dc7d6) that reproduces what you're doing here, MIRI says it's UB.